File: generated_code/GeminiPro/Python/059_SplayTree.py
class SplayTree:
    """
    A complete, self-contained Python class that implements a Splay Tree.

    This class provides a dictionary-like set interface for storing unique integer keys.
    It supports the essential methods: insert(key), delete(key), and search(key).
    The splay operation is performed on accessed nodes to move them to the root,
    optimizing for frequent access to the same elements.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Splays the node x to the root of the tree.
        This operation moves the node x to the root through a series of rotations.
        """
        while x.parent:
            p = x.parent
            g = p.parent
            if not g:  # Zig case
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-Zig case (left-left)
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-Zig case (right-right)
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-Zag case (left-right)
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-Zag case (right-left)
                self._right_rotate(p)
                self._left_rotate(g)

    def search(self, key):
        """
        Searches for a key in the tree and splays the accessed node.

        If the key is found, the corresponding node is splayed to the root.
        If the key is not found, the last non-null node accessed during the
        search is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        if last_node:
            self._splay(last_node)
        
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        Otherwise, a new node is created, inserted as in a regular BST, and
        then splayed to the root.

        Args:
            key: The integer key to insert.
        """
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                # Key already exists, splay it to the root and we're done
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        new_node = self._Node(key)
        new_node.parent = parent
        if parent is None:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
            
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        The node with the given key is first splayed to the root. Then, it is
        removed by joining its left and right subtrees. To join, the maximum
        element of the left subtree is splayed to its root and then the right
        subtree is attached as its right child.

        Args:
            key: The integer key to delete.
        """
        if not self.search(key):
            # Key not found, nothing to delete.
            # search() has already splayed the closest node.
            return
            
        # After search, the node to be deleted is at the root.
        # This is guaranteed if the key was found.
        left_subtree = self.root.left
        right_subtree = self.root.right
        
        if not left_subtree:
            # No left subtree, so the right subtree becomes the new tree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect left subtree to treat it as its own tree
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this max_node to the root of its subtree.
            # We can do this by temporarily setting self.root to the left subtree's root.
            self.root = left_subtree
            self._splay(max_node)
            
            # After splaying, self.root is max_node. We reattach the original right subtree.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/059_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for an integer set.
 * A splay tree is a self-balancing binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * It performs basic operations such as insertion, deletion, and search.
 * After a search or an access, the accessed node is moved to the root of
 * the tree using a series of rotations (splaying).
 */
public class SplayTree {

    /**
     * Inner class representing a node in the splay tree.
     */
    private class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.parent = null;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a left rotation on the given node x.
     *
     *      x               y
     *     / \             / \
     *    A   y    -->    x   C
     *       / \         / \
     *      B   C       A   B
     *
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node y.
     *
     *        y           x
     *       / \         / \
     *      x   C  -->  A   y
     *     / \             / \
     *    A   B           B   C
     *
     * @param y The node to rotate.
     */
    private void rightRotate(Node y) {
        Node x = y.left;
        if (x != null) {
            y.left = x.right;
            if (x.right != null) {
                x.right.parent = y;
            }
            x.parent = y.parent;
            if (y.parent == null) {
                this.root = x;
            } else if (y == y.parent.left) {
                y.parent.left = x;
            } else {
                y.parent.right = x;
            }
            x.right = y;
            y.parent = x;
        }
    }

    /**
     * Moves the given node to the root of the tree using splay operations.
     *
     * @param node The node to splay.
     */
    private void splay(Node node) {
        while (node.parent != null) {
            Node parent = node.parent;
            Node grandParent = parent.parent;

            if (grandParent == null) { // Zig case
                if (node == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else {
                if (node == parent.left) {
                    if (parent == grandParent.left) { // Zig-Zig case
                        rightRotate(grandParent);
                        rightRotate(parent);
                    } else { // Zig-Zag case
                        rightRotate(parent);
                        leftRotate(grandParent);
                    }
                } else {
                    if (parent == grandParent.left) { // Zig-Zag case
                        leftRotate(parent);
                        rightRotate(grandParent);
                    } else { // Zig-Zig case
                        leftRotate(grandParent);
                        leftRotate(parent);
                    }
                }
            }
        }
    }
    
    /**
     * Searches for a key in the tree. If the key is found, the corresponding
     * node is splayed to the root. If not found, the last accessed node on the
     * search path is splayed to the root.
     *
     * @param key The integer key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }

        Node current = root;
        while (current != null) {
            if (key < current.key) {
                if (current.left == null) {
                    splay(current); // Key not found, splay last accessed node
                    return null;
                }
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    splay(current); // Key not found, splay last accessed node
                    return null;
                }
                current = current.right;
            } else {
                splay(current); // Key found, splay it
                return current.key;
            }
        }
        return null; // Should be unreachable
    }

    /**
     * Inserts a key into the splay tree. If the key already exists, the existing
     * node is splayed to the root. Otherwise, the new node is inserted and then
     * splayed to the root.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        Node current = root;
        Node parent = null;
        while (current != null) {
            parent = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key already exists, splay the node and return
                splay(current);
                return;
            }
        }

        Node newNode = new Node(key);
        newNode.parent = parent;
        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }

        splay(newNode);
    }

    /**
     * Deletes a key from the splay tree.
     * The node containing the key is first splayed to the root and then removed.
     * The remaining subtrees are then merged.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        // Splay the node with the key (or the last accessed node) to the root
        if (search(key) == null) {
             // Key not in tree. search() already splayed the last accessed node.
            return;
        }

        // Now the node to delete is the root
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
            if (root != null) {
                root.parent = null;
            }
        } else {
            // Disconnect left subtree
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree
            Node maxInLeft = leftSubtree;
            while (maxInLeft.right != null) {
                maxInLeft = maxInLeft.right;
            }
            
            // Splay the maximum node to the root of the left subtree
            // We can do this by temporarily setting the root to the left subtree's root
            Node originalRoot = this.root;
            this.root = leftSubtree;
            splay(maxInLeft);

            // The new root is maxInLeft. Attach the original right subtree.
            this.root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = this.root;
            }
        }
    }
}